File: go/internal/scanner/scanner.go
// Package scanner is the native Go implementation of envdoctor's core:
// reconcile environment variables used in Go source against those defined in
// .env files. Local-first — no network, values never printed.
package scanner

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Origin is where a variable was seen.
type Origin struct {
	File string
	Line int
}

// Finding is one reported issue.
type Finding struct {
	Rule     string
	Severity string // "error" | "warning"
	Name     string
	Message  string
	Origin   Origin
}

// Result holds all findings from a scan.
type Result struct {
	Findings []Finding
}

// Errors returns only error-severity findings.
func (r Result) Errors() []Finding { return r.filter("error") }

// Warnings returns only warning-severity findings.
func (r Result) Warnings() []Finding { return r.filter("warning") }

func (r Result) filter(sev string) []Finding {
	var out []Finding
	for _, f := range r.Findings {
		if f.Severity == sev {
			out = append(out, f)
		}
	}
	return out
}

var usagePatterns = []*regexp.Regexp{
	regexp.MustCompile(`\bos\.Getenv\(\s*"([A-Za-z_]\w*)"`),
	regexp.MustCompile(`\bos\.LookupEnv\(\s*"([A-Za-z_]\w*)"`),
}

var (
	lineComment  = regexp.MustCompile(`(?m)//[^\n]*`)
	blockComment = regexp.MustCompile(`(?s)/\*.*?\*/`)
	envLine      = regexp.MustCompile(`^\s*(?:export\s+)?([A-Za-z_]\w*)\s*=(.*)$`)
)

// publicPrefixes are client-exposed env prefixes (case-sensitive, exact).
var publicPrefixes = []string{
	"NEXT_PUBLIC_",
	"VITE_",
	"REACT_APP_",
	"EXPO_PUBLIC_",
	"GATSBY_",
	"NUXT_PUBLIC_",
	"VUE_APP_",
	"PUBLIC_",
}

// secretName matches secret-looking variable names (case-insensitive).
var secretName = regexp.MustCompile(`(?i)SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH`)

// weakSecretValue matches weak/placeholder secret values (case-insensitive).
var weakSecretValue = regexp.MustCompile(`(?i)^(changeme|change_me|placeholder|x{3,}|todo|secret|password|passwd|test|example|sample|dummy|your[_-].*|<.*>|\$\{.*\})$`)

var (
	intRe   = regexp.MustCompile(`^-?\d+$`)
	floatRe = regexp.MustCompile(`^-?\d+\.\d+$`)
	urlRe   = regexp.MustCompile(`(?i)^https?://`)
)

func hasPublicPrefix(name string) bool {
	for _, p := range publicPrefixes {
		if strings.HasPrefix(name, p) {
			return true
		}
	}
	return false
}

// blankMatch replaces every non-newline rune with a space to preserve offsets.
func blankMatch(s string) string {
	var b strings.Builder
	for _, r := range s {
		if r == '\n' {
			b.WriteRune('\n')
		} else {
			b.WriteByte(' ')
		}
	}
	return b.String()
}

func stripNoise(code string) string {
	code = blockComment.ReplaceAllStringFunc(code, blankMatch)
	code = lineComment.ReplaceAllStringFunc(code, blankMatch)
	return code
}

// parseValue trims the RHS and strips one matching pair of surrounding quotes.
func parseValue(rhs string) string {
	v := strings.TrimSpace(rhs)
	if len(v) >= 2 && (v[0] == '\'' || v[0] == '"') && v[len(v)-1] == v[0] {
		v = v[1 : len(v)-1]
	}
	return v
}

// envLabel derives an environment label from a dotenv filename.
func envLabel(filename string) string {
	if filename == ".env" {
		return "default"
	}
	rest := strings.TrimPrefix(filename, ".env.")
	if rest == "local" {
		return "local"
	}
	rest = strings.TrimSuffix(rest, ".local")
	return rest
}

// InferType infers a coarse type from a value string.
func InferType(value string) string {
	if value == "" {
		return "empty"
	}
	if intRe.MatchString(value) {
		return "integer"
	}
	if floatRe.MatchString(value) {
		return "float"
	}
	if lv := strings.ToLower(value); lv == "true" || lv == "false" {
		return "boolean"
	}
	if urlRe.MatchString(value) {
		return "url"
	}
	if value[0] == '{' || value[0] == '[' {
		var js interface{}
		if json.Unmarshal([]byte(value), &js) == nil {
			return "json"
		}
	}
	return "string"
}

func compatGroup(inferred string) string {
	if inferred == "integer" || inferred == "float" {
		return "numeric"
	}
	return inferred
}

// Levenshtein computes the edit distance between two strings.
func Levenshtein(a, b string) int {
	ra, rb := []rune(a), []rune(b)
	if a == b {
		return 0
	}
	if len(ra) == 0 {
		return len(rb)
	}
	if len(rb) == 0 {
		return len(ra)
	}
	prev := make([]int, len(rb)+1)
	for j := range prev {
		prev[j] = j
	}
	for i := 1; i <= len(ra); i++ {
		cur := make([]int, len(rb)+1)
		cur[0] = i
		for j := 1; j <= len(rb); j++ {
			cost := 1
			if ra[i-1] == rb[j-1] {
				cost = 0
			}
			cur[j] = min3(prev[j]+1, cur[j-1]+1, prev[j-1]+cost)
		}
		prev = cur
	}
	return prev[len(rb)]
}

func min3(a, b, c int) int {
	m := a
	if b < m {
		m = b
	}
	if c < m {
		m = c
	}
	return m
}

// def tracks a variable's first origin and its value per environment label.
type def struct {
	origin Origin
	values map[string]string // label -> value
}

// ScanSource returns variable name -> first origin for env usage in Go source.
func ScanSource(path, content string) map[string]Origin {
	text := stripNoise(content)
	used := map[string]Origin{}
	for _, re := range usagePatterns {
		for _, m := range re.FindAllStringSubmatchIndex(text, -1) {
			name := text[m[2]:m[3]]
			if _, ok := used[name]; ok {
				continue
			}
			line := strings.Count(text[:m[0]], "\n") + 1
			used[name] = Origin{File: path, Line: line}
		}
	}
	return used
}

// envOccurrence is one definition of a key in a dotenv file.
type envOccurrence struct {
	Origin Origin
	Value  string
}

// ParseEnv returns variable name -> all occurrences (in file order) for the
// definitions in a dotenv file. Collecting every occurrence lets callers both
// reconcile against the first origin and detect in-file duplicates.
func ParseEnv(path, content string) map[string][]envOccurrence {
	defined := map[string][]envOccurrence{}
	for i, raw := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(raw)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if m := envLine.FindStringSubmatch(raw); m != nil {
			defined[m[1]] = append(defined[m[1]], envOccurrence{
				Origin: Origin{File: path, Line: i + 1},
				Value:  parseValue(m[2]),
			})
		}
	}
	return defined
}

func skipDir(name string) bool {
	switch name {
	case ".git", "vendor", "node_modules":
		return true
	}
	return false
}

func isEnvFile(base string) bool {
	return base == ".env" || (strings.HasPrefix(base, ".env.") && !strings.HasSuffix(base, ".example"))
}

// Scan reconciles .env definitions against .go source usage under root.
func Scan(root string) (Result, error) {
	defined := map[string]*def{}
	used := map[string]Origin{}
	projectLabels := map[string]bool{}
	var duplicates []Finding

	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if skipDir(info.Name()) {
				return filepath.SkipDir
			}
			return nil
		}
		base := info.Name()
		isEnv := isEnvFile(base)
		isGo := strings.HasSuffix(base, ".go")
		if !isEnv && !isGo {
			return nil
		}
		data, readErr := os.ReadFile(path)
		if readErr != nil {
			return readErr
		}
		rel, _ := filepath.Rel(root, path)
		if isEnv {
			label := envLabel(base)
			projectLabels[label] = true
			for k, occ := range ParseEnv(rel, string(data)) {
				d, ok := defined[k]
				if !ok {
					d = &def{origin: occ[0].Origin, values: map[string]string{}}
					defined[k] = d
				}
				if _, seen := d.values[label]; !seen {
					d.values[label] = occ[0].Value
				}
				// Duplicate within a single file: 2+ occurrences of a key.
				if len(occ) >= 2 {
					lines := make([]string, len(occ))
					for i, o := range occ {
						lines[i] = strconv.Itoa(o.Origin.Line)
					}
					duplicates = append(duplicates, Finding{
						Rule: "duplicates", Severity: "error", Name: k,
						Message: fmt.Sprintf("defined %d times in the same file (lines %s)",
							len(occ), strings.Join(lines, ", ")),
						Origin: occ[0].Origin,
					})
				}
			}
		} else {
			for k, v := range ScanSource(rel, string(data)) {
				if _, ok := used[k]; !ok {
					used[k] = v
				}
			}
		}
		return nil
	})
	if err != nil {
		return Result{}, err
	}

	var res Result

	// --- Errors, in rule order ---

	// undefined-in-source: used but never defined.
	for _, name := range sortedOrigins(used) {
		if _, ok := defined[name]; !ok {
			res.Findings = append(res.Findings, Finding{
				Rule: "undefined-in-source", Severity: "error", Name: name,
				Message: "used in source code but not defined in any environment file",
				Origin:  used[name],
			})
		}
	}

	// duplicates: same key defined 2+ times within a single .env file.
	sort.Slice(duplicates, func(i, j int) bool { return duplicates[i].Name < duplicates[j].Name })
	res.Findings = append(res.Findings, duplicates...)

	// public-prefix: secret-looking var exposed to client bundles.
	for _, name := range sortedDefs(defined) {
		if hasPublicPrefix(name) && secretName.MatchString(name) {
			res.Findings = append(res.Findings, Finding{
				Rule: "public-prefix", Severity: "error", Name: name,
				Message: "secret-looking variable is exposed to client bundles via a public prefix",
				Origin:  defined[name].origin,
			})
		}
	}

	// type-mismatch: incompatible inferred types across environments.
	for _, name := range sortedDefs(defined) {
		d := defined[name]
		if len(d.values) < 2 {
			continue
		}
		groups := map[string]bool{}
		for _, v := range d.values {
			t := InferType(v)
			if t == "empty" {
				continue
			}
			groups[compatGroup(t)] = true
		}
		if len(groups) >= 2 {
			res.Findings = append(res.Findings, Finding{
				Rule: "type-mismatch", Severity: "error", Name: name,
				Message: "inferred type differs across environments",
				Origin:  d.origin,
			})
		}
	}

	// --- Warnings, in rule order ---

	// unused: defined but never referenced.
	for _, name := range sortedDefs(defined) {
		if _, ok := used[name]; !ok {
			res.Findings = append(res.Findings, Finding{
				Rule: "unused", Severity: "warning", Name: name,
				Message: "defined but never referenced in source",
				Origin:  defined[name].origin,
			})
		}
	}

	// environment-diff: defined in some but not all project env labels.
	if len(projectLabels) >= 2 {
		for _, name := range sortedDefs(defined) {
			present := sortedStrings(defined[name].values)
			var absent []string
			for label := range projectLabels {
				if _, ok := defined[name].values[label]; !ok {
					absent = append(absent, label)
				}
			}
			sort.Strings(absent)
			if len(present) > 0 && len(absent) > 0 {
				res.Findings = append(res.Findings, Finding{
					Rule: "environment-diff", Severity: "warning", Name: name,
					Message: fmt.Sprintf("defined in %s but missing in %s",
						strings.Join(present, ", "), strings.Join(absent, ", ")),
					Origin: defined[name].origin,
				})
			}
		}
	}

	// weak-secret: secret-looking name with a weak or placeholder value.
	for _, name := range sortedDefs(defined) {
		if !secretName.MatchString(name) {
			continue
		}
		d := defined[name]
		value, ok := d.values[envLabel(filepath.Base(d.origin.File))]
		if !ok {
			// Fall back to any recorded value deterministically.
			for _, label := range sortedStrings(d.values) {
				value = d.values[label]
				break
			}
		}
		if value == "" || len(value) < 8 || weakSecretValue.MatchString(value) {
			res.Findings = append(res.Findings, Finding{
				Rule: "weak-secret", Severity: "warning", Name: name,
				Message: "secret-looking variable has a weak or placeholder value",
				Origin:  d.origin,
			})
		}
	}

	// typo: used-but-undefined name close to a defined name.
	definedNames := sortedDefs(defined)
	for _, u := range sortedOrigins(used) {
		if _, ok := defined[u]; ok {
			continue
		}
		bestD := ""
		bestDist := 0
		for _, dName := range definedNames {
			if dName == u {
				continue
			}
			threshold := 2
			if minInt(len(u), len(dName)) <= 4 {
				threshold = 1
			}
			dist := Levenshtein(u, dName)
			if dist <= threshold && (bestD == "" || dist < bestDist) {
				bestD, bestDist = dName, dist
			}
		}
		if bestD != "" {
			res.Findings = append(res.Findings, Finding{
				Rule: "typo", Severity: "warning", Name: u,
				Message: fmt.Sprintf("%q is not defined; did you mean %q?", u, bestD),
				Origin:  used[u],
			})
		}
	}

	return res, nil
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func sortedOrigins(m map[string]Origin) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func sortedDefs(m map[string]*def) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func sortedStrings(m map[string]string) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// DefinedByLabel maps each environment label to the set of variable names
// defined in it (across all matching .env files).
func DefinedByLabel(root string) (map[string]map[string]bool, error) {
	labels := map[string]map[string]bool{}
	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if skipDir(info.Name()) {
				return filepath.SkipDir
			}
			return nil
		}
		if !isEnvFile(info.Name()) {
			return nil
		}
		data, readErr := os.ReadFile(path)
		if readErr != nil {
			return readErr
		}
		rel, _ := filepath.Rel(root, path)
		label := envLabel(info.Name())
		set := labels[label]
		if set == nil {
			set = map[string]bool{}
			labels[label] = set
		}
		for k := range ParseEnv(rel, string(data)) {
			set[k] = true
		}
		return nil
	})
	return labels, err
}

func sortedSet(s map[string]bool) []string {
	out := make([]string, 0, len(s))
	for k := range s {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// Diff is the result of comparing two environment labels.
type Diff struct {
	OnlyInA []string `json:"onlyInA"`
	OnlyInB []string `json:"onlyInB"`
	Common  []string `json:"common"`
}

// DiffLabels compares the variable names defined in labels a and b.
func DiffLabels(root, a, b string) (Diff, error) {
	labels, err := DefinedByLabel(root)
	if err != nil {
		return Diff{}, err
	}
	da, db := labels[a], labels[b]
	d := Diff{OnlyInA: []string{}, OnlyInB: []string{}, Common: []string{}}
	for _, k := range sortedSet(da) {
		if db[k] {
			d.Common = append(d.Common, k)
		} else {
			d.OnlyInA = append(d.OnlyInA, k)
		}
	}
	for _, k := range sortedSet(db) {
		if !da[k] {
			d.OnlyInB = append(d.OnlyInB, k)
		}
	}
	return d, nil
}

// SyncLabels appends keys present in `from` but missing from `to` to to's file
// as `KEY=` placeholders. Values are never copied.
func SyncLabels(root, from, to string, dryRun bool) ([]string, error) {
	labels, err := DefinedByLabel(root)
	if err != nil {
		return nil, err
	}
	missing := []string{}
	for _, k := range sortedSet(labels[from]) {
		if !labels[to][k] {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 && !dryRun {
		target := filepath.Join(root, ".env")
		if to != "default" {
			target = filepath.Join(root, ".env."+to)
		}
		existing, _ := os.ReadFile(target)
		var b strings.Builder
		if len(existing) > 0 && existing[len(existing)-1] != '\n' {
			b.WriteString("\n")
		}
		for _, k := range missing {
			b.WriteString(k + "=\n")
		}
		f, ferr := os.OpenFile(target, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
		if ferr != nil {
			return nil, ferr
		}
		defer f.Close()
		if _, werr := f.WriteString(b.String()); werr != nil {
			return nil, werr
		}
	}
	return missing, nil
}

File: php/src/Scanner.php
<?php

declare(strict_types=1);

namespace Envdoctor;

/**
 * Core scanner: reconcile environment access in PHP source against .env
 * definitions. Local-first — no network, values never printed.
 */
final class Scanner
{
    /** @var string[] Each pattern captures the variable name in group 1. */
    private const USAGE_PATTERNS = [
        '/\bgetenv\(\s*["\']([A-Za-z_]\w*)["\']/',
        '/\$_ENV\[\s*["\']([A-Za-z_]\w*)["\']\s*\]/',
        '/\$_SERVER\[\s*["\']([A-Za-z_]\w*)["\']\s*\]/',
    ];

    private const ENV_LINE = '/^\s*(?:export\s+)?([A-Za-z_]\w*)\s*=/';

    /** @var string[] Public/client-bundle prefixes (case-sensitive). */
    private const PUBLIC_PREFIXES = [
        'NEXT_PUBLIC_', 'VITE_', 'REACT_APP_', 'EXPO_PUBLIC_',
        'GATSBY_', 'NUXT_PUBLIC_', 'VUE_APP_', 'PUBLIC_',
    ];

    private const SECRET_RE = '/SECRET|TOKEN|PASSWORD|PASSWD|PRIVATE|CREDENTIAL|API_?KEY|ACCESS_?KEY|AUTH/i';

    private const WEAK_VALUE_RE =
        '/^(changeme|change_me|placeholder|x{3,}|todo|secret|password|passwd|test|example|sample|dummy|your[_-].*|<.*>|\$\{.*\})$/i';

    private static function blank(string $s): string
    {
        return preg_replace('/[^\n]/', ' ', $s);
    }

    private static function stripNoise(string $code): string
    {
        // Block comments, then line comments (// and #).
        $code = preg_replace_callback('/\/\*.*?\*\//s', fn($m) => self::blank($m[0]), $code);
        $code = preg_replace_callback('/(?:\/\/|#)[^\n]*/', fn($m) => self::blank($m[0]), $code);

        return $code;
    }

    /** @return array<string, array{file: string, line: int}> */
    public static function scanSource(string $path, string $content): array
    {
        $text = self::stripNoise($content);
        $used = [];
        foreach (self::USAGE_PATTERNS as $pattern) {
            if (preg_match_all($pattern, $text, $matches, PREG_OFFSET_CAPTURE)) {
                foreach ($matches[1] as $m) {
                    $name = $m[0];
                    if (isset($used[$name])) {
                        continue;
                    }
                    $line = substr_count(substr($text, 0, $m[1]), "\n") + 1;
                    $used[$name] = ['file' => $path, 'line' => $line];
                }
            }
        }

        return $used;
    }

    /**
     * Parse the VALUE to the right of the first `=`: trim, then strip one pair
     * of matching surrounding quotes. Values are used ONLY for detection and
     * never appear in any output.
     */
    public static function parseValue(string $raw): string
    {
        $idx = strpos($raw, '=');
        if ($idx === false) {
            return '';
        }
        $value = trim(substr($raw, $idx + 1));
        $len = strlen($value);
        if ($len >= 2) {
            $first = $value[0];
            if (($first === '"' || $first === "'") && $value[$len - 1] === $first) {
                $value = substr($value, 1, $len - 2);
            }
        }

        return $value;
    }

    /**
     * Returns all occurrences per key: name => list of {line, value}, in order.
     *
     * @return array<string, array<int, array{line: int, value: string}>>
     */
    public static function parseEnv(string $content): array
    {
        $defined = [];
        foreach (explode("\n", $content) as $i => $raw) {
            $trimmed = trim($raw);
            if ($trimmed === '' || str_starts_with($trimmed, '#')) {
                continue;
            }
            if (preg_match(self::ENV_LINE, $raw, $m)) {
                $defined[$m[1]][] = ['line' => $i + 1, 'value' => self::parseValue($raw)];
            }
        }

        return $defined;
    }

    /** Derive the environment label from a .env filename. */
    public static function envLabel(string $filename): string
    {
        $base = basename($filename);
        if ($base === '.env') {
            return 'default';
        }
        $label = preg_replace('/^\.env\./', '', $base);
        if (str_ends_with($label, '.local')) {
            $label = substr($label, 0, -strlen('.local'));
        }

        return $label;
    }

    private static function isPublicPrefix(string $name): bool
    {
        $hasPrefix = false;
        foreach (self::PUBLIC_PREFIXES as $p) {
            if (str_starts_with($name, $p)) {
                $hasPrefix = true;
                break;
            }
        }

        return $hasPrefix && preg_match(self::SECRET_RE, $name) === 1;
    }

    /** Infer the coarse type of a value string. */
    public static function inferType(string $value): string
    {
        if ($value === '') {
            return 'empty';
        }
        if (preg_match('/^-?\d+$/', $value)) {
            return 'integer';
        }
        if (preg_match('/^-?\d+\.\d+$/', $value)) {
            return 'float';
        }
        if (preg_match('/^(true|false)$/i', $value)) {
            return 'boolean';
        }
        if (preg_match('#^https?://#', $value)) {
            return 'url';
        }
        if ($value[0] === '{' || $value[0] === '[') {
            json_decode($value);
            if (json_last_error() === JSON_ERROR_NONE) {
                return 'json';
            }
        }

        return 'string';
    }

    /** Compatibility group (integer/float collapse to numeric). */
    private static function typeGroup(string $type): string
    {
        return ($type === 'integer' || $type === 'float') ? 'numeric' : $type;
    }

    private static function isWeakSecret(string $value): bool
    {
        return $value === '' || strlen($value) < 8 || preg_match(self::WEAK_VALUE_RE, $value) === 1;
    }

    private static function levenshteinDistance(string $a, string $b): int
    {
        return levenshtein($a, $b);
    }

    /** @return string[] */
    private static function discoverFiles(string $root, string $suffix, bool $envMode): array
    {
        $out = [];
        $it = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($root, \FilesystemIterator::SKIP_DOTS)
        );
        foreach ($it as $file) {
            $name = $file->getFilename();
            $skip = false;
            foreach (['.git', 'vendor', 'node_modules'] as $bad) {
                if (str_contains($file->getPathname(), DIRECTORY_SEPARATOR . $bad . DIRECTORY_SEPARATOR)) {
                    $skip = true;
                    break;
                }
            }
            if ($skip) {
                continue;
            }
            if ($envMode) {
                $isEnv = $name === '.env' || (str_starts_with($name, '.env.') && !str_ends_with($name, '.example'));
                if ($isEnv) {
                    $out[] = $file->getPathname();
                }
            } elseif (str_ends_with($name, $suffix)) {
                $out[] = $file->getPathname();
            }
        }
        sort($out);

        return $out;
    }

    /** @return Finding[] */
    /**
     * Map each environment label to the sorted list of variable names in it.
     *
     * @return array<string, string[]>
     */
    public static function definedByLabel(string $root): array
    {
        $root = rtrim($root, DIRECTORY_SEPARATOR);
        $labels = [];
        foreach (self::discoverFiles($root, '', true) as $f) {
            $label = self::envLabel($f);
            $labels[$label] ??= [];
            foreach (array_keys(self::parseEnv((string) file_get_contents($f))) as $name) {
                if (!in_array($name, $labels[$label], true)) {
                    $labels[$label][] = $name;
                }
            }
        }

        return $labels;
    }

    /** @return array{onlyInA: string[], onlyInB: string[], common: string[]} */
    public static function diffLabels(string $root, string $a, string $b): array
    {
        $labels = self::definedByLabel($root);
        $da = $labels[$a] ?? [];
        $db = $labels[$b] ?? [];
        $sort = static function (array $x): array {
            sort($x);
            return array_values($x);
        };

        return [
            'onlyInA' => $sort(array_diff($da, $db)),
            'onlyInB' => $sort(array_diff($db, $da)),
            'common' => $sort(array_intersect($da, $db)),
        ];
    }

    /**
     * Append keys present in `from` but missing from `to` as `KEY=` placeholders.
     * Values are never copied.
     *
     * @return string[]
     */
    public static function syncLabels(string $root, string $from, string $to, bool $dryRun = false): array
    {
        $root = rtrim($root, DIRECTORY_SEPARATOR);
        $labels = self::definedByLabel($root);
        $missing = array_values(array_diff($labels[$from] ?? [], $labels[$to] ?? []));
        sort($missing);
        if ($missing && !$dryRun) {
            $target = $root . DIRECTORY_SEPARATOR . ($to === 'default' ? '.env' : ".env.$to");
            $existing = is_file($target) ? (string) file_get_contents($target) : '';
            $prefix = ($existing === '' || str_ends_with($existing, "\n")) ? '' : "\n";
            $append = $prefix . implode('', array_map(static fn($k) => "$k=\n", $missing));
            file_put_contents($target, $append, FILE_APPEND);
        }

        return $missing;
    }

    public static function scan(string $root): array
    {
        $root = rtrim($root, DIRECTORY_SEPARATOR);
        $defined = [];        // name => {file, line} (first definition wins)
        $definedValue = [];   // name => value at first definition
        $labelsOf = [];       // name => [label => value] (first value per label)
        $projectLabels = [];
        $dupFindings = [];

        foreach (self::discoverFiles($root, '', true) as $f) {
            $rel = self::rel($root, $f);
            $label = self::envLabel($f);
            if (!in_array($label, $projectLabels, true)) {
                $projectLabels[] = $label;
            }
            foreach (self::parseEnv((string) file_get_contents($f)) as $k => $defs) {
                if (count($defs) >= 2) {
                    $lines = array_map(fn($o) => $o['line'], $defs);
                    $dupFindings[] = new Finding(
                        'duplicates',
                        'error',
                        $k,
                        'defined ' . count($defs) . ' times in the same file (lines '
                            . implode(', ', $lines) . ')',
                        ['file' => $rel, 'line' => $defs[0]['line']]
                    );
                }
                // First occurrence (first file wins) counts as the definition.
                if (!isset($defined[$k])) {
                    $defined[$k] = ['file' => $rel, 'line' => $defs[0]['line']];
                    $definedValue[$k] = $defs[0]['value'];
                }
                if (!isset($labelsOf[$k][$label])) {
                    $labelsOf[$k][$label] = $defs[0]['value'];
                }
            }
        }

        $used = [];
        foreach (self::discoverFiles($root, '.php', false) as $f) {
            foreach (self::scanSource(self::rel($root, $f), (string) file_get_contents($f)) as $k => $v) {
                $used[$k] ??= $v;
            }
        }

        $errors = [];
        $warnings = [];

        // --- errors: undefined-in-source ---
        $usedNames = array_keys($used);
        sort($usedNames);
        foreach ($usedNames as $name) {
            if (!isset($defined[$name])) {
                $errors[] = new Finding(
                    'undefined-in-source',
                    'error',
                    $name,
                    'used in source code but not defined in any environment file',
                    $used[$name]
                );
            }
        }

        // --- errors: duplicates ---
        usort($dupFindings, fn($a, $b) => strcmp($a->name, $b->name));
        foreach ($dupFindings as $finding) {
            $errors[] = $finding;
        }

        $definedNames = array_keys($defined);
        sort($definedNames);

        // --- errors: public-prefix ---
        foreach ($definedNames as $name) {
            if (self::isPublicPrefix($name)) {
                $errors[] = new Finding(
                    'public-prefix',
                    'error',
                    $name,
                    'secret-looking variable is exposed to client bundles via a public prefix',
                    $defined[$name]
                );
            }
        }

        // --- errors: type-mismatch ---
        foreach ($definedNames as $name) {
            $labels = $labelsOf[$name] ?? [];
            if (count($labels) < 2) {
                continue;
            }
            $groups = [];
            foreach ($labels as $value) {
                $type = self::inferType($value);
                if ($type === 'empty') {
                    continue;
                }
                $group = self::typeGroup($type);
                if (!in_array($group, $groups, true)) {
                    $groups[] = $group;
                }
            }
            if (count($groups) >= 2) {
                $errors[] = new Finding(
                    'type-mismatch',
                    'error',
                    $name,
                    'inferred type differs across environments',
                    $defined[$name]
                );
            }
        }

        // --- warnings: unused ---
        foreach ($definedNames as $name) {
            if (!isset($used[$name])) {
                $warnings[] = new Finding(
                    'unused',
                    'warning',
                    $name,
                    'defined but never referenced in source',
                    $defined[$name]
                );
            }
        }

        // --- warnings: environment-diff ---
        if (count($projectLabels) >= 2) {
            foreach ($definedNames as $name) {
                $present = array_keys($labelsOf[$name] ?? []);
                sort($present);
                $absent = array_values(array_diff($projectLabels, $present));
                sort($absent);
                if (count($present) === 0 || count($absent) === 0) {
                    continue;
                }
                $warnings[] = new Finding(
                    'environment-diff',
                    'warning',
                    $name,
                    'defined in ' . implode(', ', $present) . ' but missing in ' . implode(', ', $absent),
                    $defined[$name]
                );
            }
        }

        // --- warnings: weak-secret ---
        foreach ($definedNames as $name) {
            if (preg_match(self::SECRET_RE, $name) !== 1) {
                continue;
            }
            if (!self::isWeakSecret((string) ($definedValue[$name] ?? ''))) {
                continue;
            }
            $warnings[] = new Finding(
                'weak-secret',
                'warning',
                $name,
                'secret-looking variable has a weak or placeholder value',
                $defined[$name]
            );
        }

        // --- warnings: typo ---
        foreach ($usedNames as $u) {
            if (isset($defined[$u])) {
                continue;
            }
            $best = null;
            $bestDist = null;
            foreach ($definedNames as $d) {
                if ($d === $u) {
                    continue;
                }
                $limit = min(strlen($u), strlen($d)) <= 4 ? 1 : 2;
                $dist = self::levenshteinDistance($u, $d);
                if ($dist > $limit) {
                    continue;
                }
                if ($best === null || $dist < $bestDist || ($dist === $bestDist && strcmp($d, $best) < 0)) {
                    $best = $d;
                    $bestDist = $dist;
                }
            }
            if ($best === null) {
                continue;
            }
            $warnings[] = new Finding(
                'typo',
                'warning',
                $u,
                '"' . $u . '" is not defined; did you mean "' . $best . '"?',
                $used[$u]
            );
        }

        return array_merge($errors, $warnings);
    }

    /**
     * Serialize findings to the shared JSON shape. Values never appear.
     *
     * @param Finding[] $findings
     */
    public static function toJsonArray(array $findings): string
    {
        $out = [];
        foreach ($findings as $f) {
            $out[] = [
                'rule' => $f->rule,
                'severity' => $f->severity,
                'name' => $f->name,
                'message' => $f->message,
                'file' => $f->origin['file'] ?? null,
                'line' => $f->origin['line'] ?? null,
            ];
        }

        return json_encode($out, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
    }

    private static function rel(string $root, string $path): string
    {
        return ltrim(substr($path, strlen($root)), DIRECTORY_SEPARATOR);
    }
}

/** One reported issue. */
final class Finding
{
    /** @param array{file: string, line: int} $origin */
    public function __construct(
        public string $rule,
        public string $severity,
        public string $name,
        public string $message,
        public array $origin,
    ) {
    }
}
